python/lsst/daf/butler/core/serverModels.py

```diff
@@ -0,0 +1,107 @@
+# This file is part of daf_butler.
+#
+# Developed for the LSST Data Management System.
+# This product includes software developed by the LSST Project
+# (http://www.lsst.org).
+# See the COPYRIGHT file at the top-level directory of this distribution
+# for details of code ownership.
+#
+# This program is free software: you can redistribute it and/or modify
+# it under the terms of the GNU General Public License as published by
+# the Free Software Foundation, either version 3 of the License, or
+# (at your option) any later version.
+#
+# This program is distributed in the hope that it will be useful,
+# but WITHOUT ANY WARRANTY; without even the implied warranty of
+# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+# GNU General Public License for more details.
+#
+# You should have received a copy of the GNU General Public License
+# along with this program.  If not, see <http://www.gnu.org/licenses/>.
+
+from __future__ import annotations
+
+__all__ = (
+    "QueryDatasetsModel",
+    "ExpressionQueryParameter",
+)
+
+"""Models used for client/server communication."""
+
+from typing import (
+    Any,
+    List,
+    Mapping,
+    Optional,
+    Union,
+)
+
+import re
+
+from pydantic import BaseModel
+
+from .utils import iterable, globToRegex
+
+# DataId and kwargs can only take limited types
+DataIdValues = Union[str, int, float]
+DataId = Mapping[str, DataIdValues]
+
+
+class ExpressionQueryParameter(BaseModel):
+    """Represents a specification for an expression query."""
+
+    regex: Optional[List[str]] = None
+    """List of regular expression strings."""
+
+    glob: Optional[List[str]] = None
+    """List of dataset type names or globs."""
```

Comment:
I'm a bit surprised to see globs mentioned at all here; I had considered those to be purely part of the CLI.

Reply:
I'm starting to get concerned that you are reviewing an old version of this pull request somehow.

Anyhow, yes, I called it `glob` rather than `strings` because when I was using the test server directly (and not as part of the RemoteRegistry) it was easier to be able to say `cal*` to do the wildcard matching. There will be some users of the server who aren't using Python.

Reply:
Thinking about this a bit more, maybe we make `CollectionSearch.fromExpression` (and similar) responsible for identifying glob-like strings and turning them into regexes.  It's already responsible for standardizing a bunch of other kinds of inputs, and that'd let us use the globs in the Python API all over the place.